subsys/pm/policy.c

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright (c) 2018 Intel Corporation.
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <pm/pm.h>
+#include <pm/policy.h>
+#include <spinlock.h>
+#include <sys_clock.h>
+#include <sys/__assert.h>
+#include <sys/time_units.h>
+#include <sys/atomic.h>
+
+/** State disallow reference counting */
+static atomic_t state_disallow_cnt[PM_STATE_COUNT];
+/** Callback to notify when state allowed status changes */
+static pm_policy_state_allowed_changed_cb_t allowed_changed_cb;
+
+/** Lock to synchronize access to the latency request list. */
+static struct k_spinlock latency_lock;
+/** List of maximum latency requests. */
+static sys_slist_t latency_reqs;
+/** Maximum CPU latency in ticks */
+static int32_t max_latency_ticks = K_TICKS_FOREVER;
+/** Callback to notify when maximum latency changes. */
+static pm_policy_latency_changed_cb_t latency_changed_cb;
+
+/** @brief Update maximum allowed latency. */
+static void update_max_latency(void)
+{
+	int32_t new_max_latency_ticks = K_TICKS_FOREVER;
+	struct pm_policy_latency_request *req;
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&latency_reqs, req, node) {
+		if ((new_max_latency_ticks == K_TICKS_FOREVER) ||
+		    ((int32_t)req->value < new_max_latency_ticks)) {
+			new_max_latency_ticks = (int32_t)req->value;
+		}
+	}
+
+	if ((latency_changed_cb != NULL) &&
+	    (max_latency_ticks != new_max_latency_ticks)) {
+		int32_t latency_us;
+
+		if (new_max_latency_ticks == K_TICKS_FOREVER) {
+			latency_us = SYS_FOREVER_US;
+		} else {
+			latency_us = (int32_t)k_ticks_to_us_ceil32(new_max_latency_ticks);
+		}
+
+		latency_changed_cb(latency_us);
```

Comment:
User-provided callback called with spinlock held might be a recipe for lockup.

Reply:
We can't offer safety guarantee otherwise: another user may call for a change before this completed.